### DISCUSSION TITLE
Consuming in Rust

### DISCUSSION THREAD

**Discussion by justinstroud-campmeeker** [aheber/tree-sitter-sfapex#42] [Q&A]
Howdy:

I am trying to get this grammar to work in Rust, and I'm having a heck of a time. I can get the old tree-sitter-apex one to work (and some others that I've tried.) Just not this one. I'm getting a pretty inscrutable error from CC:
`/home/jstroud/repos/rust/crustforce/crusforce-apex/src/main.rs:11: undefined reference to `tree_sitter_sfapex'`

Thoughts? Smashing ideas? Smash my computer? I'm using a physical AlmaLinux box for this, FYI.

**Comment by aheber** [comment]
@justinstroudbah I don't know much about Rust but I can take a few guesses.

This project is actually four different parsers; Apex, SOQL, SOSL, and SF Logs. I would focus on trying to consume the `./apex` directory if you aren't already. The `parser.c` file should exist in there. I don't know if you can consume that directly or if you'd need a binding file.

I have a backlog item to get the binding put together for Rust and get this published for Cargo but haven't taken the time to learn and do it. Being a multi-parser project brings some interesting challenges.

One place you might want to look would be https://github.com/xixiaofinland/afmt, @xixiaofinland has been using this library in Rust and if it isn't obvious in his source code I imagine he'd be willing to pass along any tips or tricks.

If you do get it working, please feel welcome to come and add that knowledge here for the next person that might need it.

**Comment by xixiaofinland** [comment]
It took a bit of time to find the right way, but once it's set it's straightforward. 
https://github.com/xixiaofinland/afmt/blob/a72c37de8151566f8e660de2812e63aa475f0a76/src/context.rs#L84
https://github.com/xixiaofinland/afmt/blob/main/doc/Technical.md#update-parser-version

**Comment by xixiaofinland** [comment]
I realize that you might be missing this file. 
https://github.com/xixiaofinland/afmt/blob/main/build.rs

**Comment by aheber** [comment]
@justinstroudbah I've recently started publishing this projects as a create.

https://crates.io/crates/tree-sitter-sfapex
